design-proposals/golden-image-delivery-and-update-pipeline.md

```diff
@@ -0,0 +1,272 @@
+
+## HCO: Integration with SSP Operator
+
+The HCO is responsible for setting VolumeImportPoller sources on the SSP operator’s CR. These volume sources are the locations used by the VolumeImportPollers to populate PVCs for the common templates.
+
+
+## OCP UI: Integration with VolumeReferences
+
+The UI currently detects if a template has a boot source by checking if the DataVolumeTemplate’s source PVC is present. The UI needs to now be capable of checking if a DataVolumeTemplate’s source VolumeReference has phase: linked as well in order to determine if a boot source is set.
+
+## Scalability
+
+- Volume import pollers have the potential to exhaust storage resources. This can be mitigated through a garbage collection policy for out-of-date disk volumes. 
+- A large amount of import pollers triggering at the exact same interval could impact network resources during the import. This can be mitigated by introducing a jitter factor into the recurring poll interval.
+
+## Update/Rollback Compatibility
+
+### Transitioning the CommonTemplates to VolumeReferences
+
+SSP operator must make image pipelines an opt-in feature to ensure existing boot images are not overwritten.
+
+To handle the default (non auto update/delivery) use case during update, the SSP operator will create corresponding VolumeReferences for every reference found in the common templates being deployed. The default VolumeReference will point to the PVC name used before VolumeReferences were introduced into the common templates.
+
+Here's an example of how this works.
+
+**Before Update:** If the common templates had a DataVolumeTemplate that looked like the following before VolumeReferences were introduced.
+```yaml
+dataVolumeTemplates:
+- apiVersion: cdi.kubevirt.io/v1beta1
+  kind: DataVolume
+  metadata:
+    name: fedora
+  spec:
+    pvc:
+      accessModes:
+        - ReadWriteMany
+      resources:
+        requests:
+          storage: 30Gi
+    source:
+      pvc:
+        name: fedora
+        namespace: golden-images
+```
+
+
+**After Update:** The common template would have a DataVolumeTemplate that looks like the following after the update
+```yaml
+dataVolumeTemplates:
+- apiVersion: cdi.kubevirt.io/v1beta1
+  kind: DataVolume
+  metadata:
+    name: fedora
+  spec:
+    pvc:
+      accessModes:
+        - ReadWriteMany
+      resources:
+        requests:
+          storage: 30Gi
+    source:
+      volumeReference:
+        name: fedora
+        namespace: golden-images
+
+```
+
+**After Update:** And the SSP Operator would create a VolumeReference that pointed to the old PVC location previously referenced before the update.
+```yaml
+apiVersion: cdi.kubevirt.io/v1beta1
+kind: VolumeReference
+metadata:
+  name: fedora
+spec:
+  source:
```

Comment:
Do we need `source` and/or `pvc`?  Maybe `spec.source.namespace` or `spec.pvc.namespace`

Reply:
> Do we need source and/or pvc? Maybe spec.source.namespace or spec.pvc.namespace

if we want to leave the door open for multiple sources in the future, then i think we'd need it structured like this. If we can commit to only using VolumeReferences for PVCs, then source and pvc are irrelevant. 

This was designed from the perspective of keeping that door open to multiple sources.

Reply:
Agree if supporting multiple sources